MC Mirror flag clarification (#442)
minio/docs: source/reference/minio-mc/mc-mirror.rst

.. start-mc-mirror-desc

The :mc:`mc mirror` command synchronizes content to MinIO deployment, similar to the ``rsync`` utility.
:mc:`mc mirror` supports filesystems, MinIO deployments, and other S3-compatible hosts as the synchronization source.

.. end-mc-mirror-desc

.. tab-set::

   .. tab-item:: EXAMPLE

      The following command synchronizes content from a local filesystem directory to the ``mydata`` bucket on the ``myminio`` MinIO deployment.

      .. code-block:: shell
         :class: copyable

         mc mirror --watch ~/mydata myminio/mydata

>       The command "watches" for files added or removed on the local filesystem and synchronizes those operations to MinIO until explicitly terminated from the terminal. 

Reviewer: Nit - terminated from the terminal is a little redundant (and if someone forked the process, that would require different steps)

It should be sufficient to state "until explicitly terminated"

>       `--watch` does not update files changed on the local filesystem to MinIO (see `--overwrite`).

Reviewer: If you want to make these linkable, you can use `mc-cmd` - for example:

```suggestion
      `--watch` does not update files changed on the local filesystem to MinIO (see :mc-cmd:`~mc mirror`--overwrite`).
```

The tild automatically truncates everything up to the last argument, so it *should* output a link to the `--overwrite` anchor.

.. mc-cmd:: --encrypt-key
   

>    Encrypt or decrypt objects using server-side encryption with client-specified keys. 

Reviewer: So we don't have to answer this *now*, but it might be worth asking (@aead maybe?) whether this is a blind en/de crypt

How do we handle mixed encryption states? e.g. if the source has both encrypted and unencrypted objects (and for the sake of sanity lets assume that all encrypted objects use the same key), what happens on the destination? Do we decrypt the encrypted objects and encrypt the decrypted objects?

Author: Added https://github.com/minio/docs/issues/443 to cover this later.

   
   Overwrites object(s) on the :mc-cmd:`~mc mirror TARGET`.

>    For example, object A exists on both the Source and the Target.
>    Object A on the source changes.
>    Running `mc mirror --overwrite` replaces the version on the target with the version from the Source.
> 

Reviewer: We should restructure this in context of `--overwrite` already running, and then a change occurring:

```suggestion
   For example, consider an active `mc mirror --overwrite` synchronizing content from source to destination.
   
   If an object on the source changes, `mc mirror` synchronizes and overwrites any matching file on the destination.
   
   Without --overwrite, <behavior>

```

The inclusion of *version* here also might need some care. Buckets can also be unversioned, in which case I assume it's just a straight overwrite.

.. mc-cmd:: --preserve, a
   

   Preserve file system attributes and bucket policy rules of the :mc-cmd:`~mc mirror SOURCE` on the :mc-cmd:`~mc mirror TARGET`.

.. mc-cmd:: --region
   

   Specify the ``string`` region when creating new bucket(s) on the target. 

   Defaults to ``"us-east-1"``.

.. mc-cmd:: --remove
   

   Removes object(s) on the Target that do not exist on the Source. 

   For example, objects A, B, and C exist on Source.
   Objects C, D, and E exist on Target.

   When running `mc mirror --remove`, objects A and B copy to Target and objects D and E are removed from the Target.

>    C already exists on both, so does not change on either.

Reviewer: Might clarify here that C must be identical on source and destination. If the C on the destination differs - actually I'm not sure what would happen there - might be worth testing. With --overwrite we would overwrite, but if we are just checking for filenames and not checksums, the behavior is less clear.

>    After the action, only objects A, B, and C exist on both teh Source and the Target.

Reviewer: ```suggestion
   After the action, only objects A, B, and C exist on both the Source and the Target.
```

   Use the `--remove` flag to keep the same set of objects on both Source and Target.

.. mc-cmd:: storage-class, sc
   

   Set the storage class for the new object(s) on the :mc-cmd:`~mc mirror TARGET`. 
         
   See the Amazon documentation on :aws-docs:`Storage Classes <AmazonS3/latest/dev/storage-class-intro.html>` for more information on S3 storage classses.

.. mc-cmd:: --watch, w
   

   Continuously monitor the :mc-cmd:`~mc mirror SOURCE` path and synchronize `PUT` and `DELETE` changes from Source to Target.

   Defaults to ``0`` (all objects).

   For example, object A and B exist on the watched Source.
   Objects A, B, and C exist on the watched Target.
   

>    An action adds object D to Source and removes object B.

Reviewer: "An action" feels too generic. A client is going to perform a PUT write op. If it's a local filesystem it would be a write operation.

Maybe "A client writes object D to the source"

   After the action, objects A and D exist on the Source.
   Objects A, C, and D exist on the Target.
   

>    Use `--watch` flag to mirror objects from Source to Target, where the Target may also have additional objects not present on the Source.

Reviewer: Should this be at the top? It's a good modifier to the initial graf to set the context for usage.

   
   Overwrites object(s) on the :mc-cmd:`~mc mirror TARGET`.

   For example, consider an active ``mc mirror --overwrite`` synchronizing content from Source to Destination.

   If an object on Source changes, ``my mirror --overwrite`` synchronizes and overwrites any matching file on Destination.

>    Without ``--overwrite``, if an object already exists on the Destination, the mirror process fails for the object that exists in both locations.

Reviewer: Nit - can simplify

```suggestion
   Without ``--overwrite``, if an object already exists on the Destination, the mirror process fails to synchronize that object.
```

Curiosity - do we fail and continue, or just fail?

Author: Fail and continue.

.. mc-cmd:: --preserve, a
   

   Preserve file system attributes and bucket policy rules of the :mc-cmd:`~mc mirror SOURCE` on the :mc-cmd:`~mc mirror TARGET`.

.. mc-cmd:: --region
   

   Specify the ``string`` region when creating new bucket(s) on the target. 

   Defaults to ``"us-east-1"``.

.. mc-cmd:: --remove
   

   Removes object(s) on the Target that do not exist on the Source. 

   For example, objects A, B, and C exist on Source.
   Objects C, D, and E exist on Target.

   When running `mc mirror --remove`, objects A and B copy to Target and objects D and E are removed from the Target.

>    C already exists on both, so does not change on either. 

Reviewer: Might need to be more specific here - it's all about namespaces, when the expectation might be CRC/hash. So we're only looking to see if the set of namespaces on Source and Destination differ, and trimming the Destination to match.

An object named C exists on both source and destination. `mc mirrror --remove` only checks that "C" exists and does not verify that the contents match.

Might need to verify that is true if you haven't poked it already. I would assume so, because that would make the need for overwrite more clear

Author: Adding a paragraph that points to the `--overwrite` flag.

   After the action, only objects A, B, and C exist on both the Source and the Target.

   Use the `--remove` flag to keep the same set of objects on both Source and Target.

>    Use :mc-cmd:`~mc mirror --overwrite` or :mc-cmd:`~mc mirror --watch` if object C differs between Source and Target and the object on Source should be on both.

Reviewer: Nit: simplify

```suggestion
   Use :mc-cmd:`~mc mirror --overwrite` or :mc-cmd:`~mc mirror --watch` to overwrite objects on the Destination if they already exist.
```

.. mc-cmd:: storage-class, sc
   

   Set the storage class for the new object(s) on the :mc-cmd:`~mc mirror TARGET`. 
         
   See the Amazon documentation on :aws-docs:`Storage Classes <AmazonS3/latest/dev/storage-class-intro.html>` for more information on S3 storage classses.

.. mc-cmd:: --watch, w
   

   Use ``--watch`` flag to mirror objects from Source to Target, where the Target may also have additional objects not present on the Source.

   Defaults to ``0`` (all objects).

   For example, object A and B exist on the watched Source.
   Objects A, B, and C exist on the watched Target.
   
   A client writes object D to Source and removes object B.

   After the operation, objects A and D exist on the Source.
   Objects A, C, and D exist on the Target.

   :mc-cmd:`~mc mirror --watch` implies the :mc-cmd:`~mc mirror --overwrite` flag.

>    While watching the Source, files changed on Source also update on the Target.

Reviewer: This feels out of place here - should it be up in 221? We could break it up into three points:

- --watch continuously synchronizes files from source to destination
- The destination may have files that do not exist on the source
- --watch overwrites objects on the Destination if they match on the Source

   
   Overwrites object(s) on the :mc-cmd:`~mc mirror TARGET`.

   For example, consider an active ``mc mirror --overwrite`` synchronizing content from Source to Destination.

   If an object on Source changes, ``my mirror --overwrite`` synchronizes and overwrites any matching file on Destination.

   Without ``--overwrite``, if an object already exists on the Destination, the mirror process fails to synchronize that object.

>    MinIO logs an error and continues to synchronize other objects.

Author: Well that seems obvious.

>    MinIO logs an error and continues to synchronize other objects.

Author: That is, it seems obvious that it should be `mc mirror` here.

.. mc-cmd:: --preserve, a
   

   Preserve file system attributes and bucket policy rules of the :mc-cmd:`~mc mirror SOURCE` on the :mc-cmd:`~mc mirror TARGET`.

.. mc-cmd:: --region
   

   Specify the ``string`` region when creating new bucket(s) on the target. 

   Defaults to ``"us-east-1"``.

.. mc-cmd:: --remove
   

   Removes object(s) on the Target that do not exist on the Source. 

   Use the `--remove` flag to have the same list of objects on both Source and Target.

   For example, objects A, B, and C exist on Source.
   Objects C, D, and E exist on Target.

   When running ``mc mirror --remove``, objects A and B copy to Target and objects D and E are removed from Target.
   Since an object C already exists on both, nothing moves from Source to Target. 

   After the action, only objects A, B, and C exist on both the Source and the Target.

   Note: MinIO does not verify that the contents of object C are the same on both Source and Target, only that an object called `C` exists on both.

>    To make object C on Target the same as object C on Source, use  :mc-cmd:`~mc mirror --overwrite` or :mc-cmd:`~mc mirror --watch`.

Reviewer: ```suggestion
   To ensure objects on the Source and Destination match both names *and* content, use  :mc-cmd:`~mc mirror --overwrite` or :mc-cmd:`~mc mirror --watch`.
```

Author: Using `Target` instead of `Destination` to match the rest of this section.
Otherwise, 👍
